### PR TITLE
List restriction to base

### PR DESCRIPTION
**Pull request by postrstk**
For schema:
```xml
<xsd:simpleType name=dArrayType>
	<xsd:list itemType=xsd:double/>
</xsd:simpleType>

<xsd:simpleType name=d3ArrayType>
	<xsd:restriction base=test:dArrayType>
		<xsd:length value=3/>
	</xsd:restriction>
</xsd:simpleType>
```
when generating code, `kdwsdl2cpp` skips the call to the `entries()` method for `DArrayType`.
For example, `void D3ArrayType::setValue (const DArrayType & value)` instead of `"value.entries().length ()"` is generated by `"value.length()"`
```c++
void D3ArrayType::setValue( const DArrayType& value )
{
    bool rangeOk = true;

    rangeOk = rangeOk && (value.length() == 3);

    if (!rangeOk)
        qDebug( "Invalid range in D3ArrayType::setValue()" );

    mValue = value;
}
```

### CODE REVIEW

File: unittests/unittests.pro [hunk 1]
```diff
@@ -37,16 +37,16 @@ SUBDIRS = \
   encapsecurity \
   prefix_wsdl \
   vidyo \
+  list_restriction \
   empty_list_wsdl \
-  onvif_org_event \
```
Comment (SudakovIvan) [inline unittests/unittests.pro]: Why onvif_org_event was removed?

File: unittests/unittests.pro [hunk 2]
```diff
@@ -37,16 +37,16 @@ SUBDIRS = \
...
-  empty_element_wsdl \
   fault_namespace \
+  empty_element_wsdl \
   enzo \
   date_example \
   dv_terminalauth \
   test_calc \
   ws_addressing_support \
-  ws_usernametoken_support
```
Comment (SudakovIvan) [inline unittests/unittests.pro]: Why was "ws_usernametoken_support" removed?

File: unittests/unittests.pro [hunk 3]
```diff
@@ -37,16 +37,16 @@ SUBDIRS = \
...
+  default_attribute_value_wsdl
```
Comment (SudakovIvan) [inline unittests/unittests.pro]: These tests only exist in the Rec-SPb-ETU fork

Comment (postrstk) [reply unittests/unittests.pro]: Mistake.
Left tests from KDAB/KDSOAP + "list_restriction"

File: kdwsdl2cpp/CMakeLists.txt [hunk 4]
```diff
@@ -59,16 +59,19 @@ set(SOURCES
...
+
+add_library(kdwsdl2cpp_lib STATIC ${SOURCES_LIB})
+add_executable(kdwsdl2cpp ${SOURCES_EXE})
+
+target_link_libraries(kdwsdl2cpp kdwsdl2cpp_lib ${QT_LIBRARIES})
 
 install(TARGETS kdwsdl2cpp EXPORT "KDSoapTargets"
-  RUNTIME DESTINATION ${INSTALL_RUNTIME_DIR}
-  BUNDLE DESTINATION .
+  DESTINATION ${INSTALL_RUNTIME_DIR}/${PROJECT_NAME_FOR_INSTALL}
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/CMakeLists.txt]: What is PROJECT_NAME_FOR_INSTALL?

File: kdwsdl2cpp/CMakeLists.txt [hunk 5]
```diff
@@ -59,16 +59,19 @@ set(SOURCES
...
-target_link_libraries(kdwsdl2cpp ${QT_LIBRARIES})
+set(SOURCES_EXE src/main.cpp)
+
+add_library(kdwsdl2cpp_lib STATIC ${SOURCES_LIB})
+add_executable(kdwsdl2cpp ${SOURCES_EXE})
+
+target_link_libraries(kdwsdl2cpp kdwsdl2cpp_lib ${QT_LIBRARIES})
 
 install(TARGETS kdwsdl2cpp EXPORT "KDSoapTargets"
-  RUNTIME DESTINATION ${INSTALL_RUNTIME_DIR}
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/CMakeLists.txt]: why did you remove RUNTIME?

File: kdwsdl2cpp/src/converter_simpletype.cpp [hunk 6]
```diff
@@ -466,8 +468,19 @@ static QString escapeRegExp(const QString &str)
...
-static KODE::Code createRangeCheckCode(const XSD::SimpleType *type, const QString &baseTypeName, const QString &variableName, KODE::Class &parentClass)
+static KODE::Code createRangeCheckCode(const XSD::SimpleType *type, const QString &baseTypeName, const QString &variableName, KODE::Class &parentClass, const XSD::SimpleType &baseSimpleType)
 {
+    QString extendedVariableName = variableName;
+
+    if (!baseSimpleType.isNull()) {
+        if (baseSimpleType.subType() == XSD::SimpleType::SubType::TypeList) {
+            extendedVariableName += ".entries()";
+        }
+        else {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_simpletype.cpp]: coding style: join with previous line

File: unittests/list_restriction/test_list_restriction.cpp [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+
+#include <converter.h>
+#include <creator.h>
+#include <settings.h>
+
+#include <compiler.h>
+//#include "wsdl_import_definition.h"
+#include "httpserver_p.h"
+
+#include <QtTest/QtTest>
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: This includes all of QtTest and all of QtCore, please use `QTest` instead

File: unittests/list_restriction/test_list_restriction.cpp [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+#include <QEventLoop>
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: remove, not used

File: unittests/list_restriction/test_list_restriction.cpp [hunk 9]
```diff
@@ -0,0 +1,358 @@
+#include <wsdl.h>
+
+#include <fileprovider.h>
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: unused?

File: unittests/list_restriction/test_list_restriction.cpp [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+                                             "</definitions>");
+        QXmlInputSource source;
+        source.setData(srcXml);
+        QXmlSimpleReader reader;
+        reader.setFeature(QLatin1String("http://xml.org/sax/features/namespace-prefixes"), true);
+
+        QString errorMsg;
+        int errorLine, errorCol;
+        QDomDocument doc;
+        doc.setContent(&source, &reader, &errorMsg, &errorLine, &errorCol);
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: wrap in `QVERIFY()`

File: unittests/list_restriction/test_list_restriction.cpp [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+            KWSDL::WSDL wsdl;
+            wsdl.setDefinitions(definitions);
+            wsdl.setNamespaceManager(namespaceManager);
+            KWSDL::Converter converter;
+            converter.setWSDL(wsdl);
+
+            converter.convert();
+            foreach (auto elem, converter.classes()) {
+                if(elem.name() == QString("TEST__D3ArrayType")) {
+                    foreach(auto e_f,elem.functions()) {
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: missing space after comma, and after foreach

File: unittests/list_restriction/test_list_restriction.cpp [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+                        if(e_f.name() == QString("setValue")) {
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: missing space after if

File: unittests/list_restriction/test_list_restriction.cpp [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+            KODE::Code::setDefaultIndentation(4);
+            KWSDL::WSDL wsdl;
+            wsdl.setDefinitions(definitions);
+            wsdl.setNamespaceManager(namespaceManager);
+            KWSDL::Converter converter;
+            converter.setWSDL(wsdl);
+
+            converter.convert();
+            foreach (auto elem, converter.classes()) {
+                if(elem.name() == QString("TEST__D3ArrayType")) {
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: missing space after if

File: unittests/list_restriction/test_list_restriction.cpp [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+        NSManager namespaceManager;
+        namespaceManager.setPrefix(QLatin1String("xml"), NSManager::xmlNamespace());
+
+        MessageHandler messageHandler;
+        ParserContext context;
+        context.setNamespaceManager(&namespaceManager);
+        context.setMessageHandler(&messageHandler);
+
+        KWSDL::Definitions definitions;
+        if (definitions.loadXML(&context, element)) {
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: `QVERIFY()` rather than `if()`

File: unittests/list_restriction/test_list_restriction.cpp [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+            definitions.fixUpDefinitions(/*&context, element*/);
+            KODE::Code::setDefaultIndentation(4);
+            KWSDL::WSDL wsdl;
+            wsdl.setDefinitions(definitions);
+            wsdl.setNamespaceManager(namespaceManager);
+            KWSDL::Converter converter;
+            converter.setWSDL(wsdl);
+
+            converter.convert();
+            foreach (auto elem, converter.classes()) {
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: All this test, until this point, is very similar to the previous one.

You could extract a helper function for all three tests, which takes XML as input, and returns a list of classes as output?
With qtestlib that's a bit tricky (QVERIFY/QCOMPARE inside the helper function requires that it returns void), there are several solutions. One solution would be to take `Class::List &` as second argument to the function (and assign to that parameter at the end of the function).

File: unittests/list_restriction/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,5 @@
+project(list_restriction)
+
+set(list_restriction_SRCS test_list_restriction.cpp)
+set(EXTRA_LIBS kdwsdl2cpp_lib)
```
Comment (dfaure-kdab) [inline unittests/list_restriction/CMakeLists.txt]: Unittests for the kdwsdl2cpp code directly, that's interesting. I like the idea, even though in this particular case, the standard way would have been to save those wsdl documents as actual files, running kdwsdl2cpp on them and simply checking that the result compiles, as we do in many other unittesets :-)

But I like it, it can be useful for other bugs that require direct unittests, i.e. those where "it compiles" wouldn't be enough anyway.

File: kdwsdl2cpp/CMakeLists.txt (hunk 5 above)
Comment (postrstk) [reply kdwsdl2cpp/CMakeLists.txt]: These are changes from another branch. They are not needed here.

File: unittests/list_restriction/test_list_restriction.cpp [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+#include <parsercontext.h>
+
+#include <converter.h>
+#include <creator.h>
+#include <settings.h>
+
+#include <compiler.h>
+#include "httpserver_p.h"
+
+#include <QtTest>
```
Comment (dfaure-kdab) [inline unittests/list_restriction/test_list_restriction.cpp]: This includes all of QtTest *and* all of QtCore.

I'll change it to <QTest> instead (and add whatever is missing)

Comment (dfaure-kdab) [reply unittests/list_restriction/test_list_restriction.cpp]: (you forgot to remove the 't' so this is still a full-module-include)
